src/ocean/math/WideInt.d

```diff
@@ -0,0 +1,474 @@
+
+    public size_t digits ( )
+    {
+        WideInt copy = *this;
+        size_t count;
+        while (copy != 0)
+        {
+            copy.divideBy(10);
+            ++count;
+        }
+        return count;
+    }
+
+    /***************************************************************************
+
+        Inefficient allocating string conversion useful for tests and
+        prototyping.
+
+        Returns:
+            string representation
+
+    ***************************************************************************/
+
+    public istring toString ( )
+    {
+        mstring result;
+        this.toString((cstring s) { result ~= s; });
+        return assumeUnique(result);
+    }
+
+    unittest
+    {
+        WideInt num = 42;
+        test!("==")(num.toString(), "42");
+
+        num.payload[2] = 1;
+        test!("==")(num.toString(), "18446744073709551658");
+    }
+
+    /***************************************************************************
+
+        Sink based string conversion
+
+        Params:
+            sink = delegate to call with resulting string
+
+    ***************************************************************************/
+
+    public void toString ( void delegate (cstring) sink )
+    {
+        auto n = this.digits();
+        static mstring buffer;
+        buffer.length = n;
+        enableStomping(buffer);
+
+        WideInt copy = *this;
+        for (ptrdiff_t idx = n-1; idx >= 0; --idx)
+        {
+            auto remainder = copy.divideBy(10);
+            buffer[idx] = cast(char) ('0' + remainder);
+        }
+
+        sink(buffer);
+    }
+
+    /***************************************************************************
+
+        Enables assignment from a plain integer type
+
+        Params:
+            rhs = value to assign
+
+    ***************************************************************************/
+
+    public void opAssign ( ulong rhs )
+    {
+        this.payload[] = 0;
+        this.payload[0] = cast(uint) (rhs & uint.max);
+        this.payload[1] = cast(uint) (rhs >> 32);
+    }
+
+    unittest
+    {
+        WideInt i;
+        i = 42;
+        test!("==")(i.payload[0], 42);
+        i = ulong.max;
+        test!("==")(i.payload[0], uint.max);
+        test!("==")(i.payload[1], uint.max);
+    }
+
+    unittest
+    {
+        WideInt a = 42;
+        WideInt b = 43;
+        a = b;
+        test!("==")(a, 43);
+    }
+
+    /***************************************************************************
+
+        Enables equality comparison with a plain integer type
+
+        Params:
+            rhs = value to compare to
+
+    ***************************************************************************/
+
+    public equals_t opEquals ( ulong rhs )
+    {
+        if (this.payload[0] != (rhs & uint.max))
+            return false;
+
+        if (this.payload[1] != (rhs >> 32))
+            return false;
```

Comment:
This is where I miss the opportunity to return a `uint[2]` from some helper function.  We could encapsulate everything needed to assign from `ulong`, and also generate a single point of comparison for `this.payload[0 .. 2]` ... :-P